numpy/core/overrides.py

```diff
@@ -44,43 +55,75 @@ def get_overloaded_types_and_args(relevant_args):
     return overloaded_types, overloaded_args
 
 
-def array_function_override(overloaded_args, func, types, args, kwargs):
-    """Call __array_function__ implementations."""
+def array_function_implementation_or_override(
+        implementation_func, api_func, dispatcher, args, kwargs):
+    """Implement a function with checks for __array_function__ overrides.
+
+    Arguments
+    ---------
+    implementation_func : function
+        Function that implements the operation on NumPy array without
+        overrides when called like `implementation_func(*args, **kwargs)`.
+    api_func : function
+        Function exposed by NumPy's public API  on which overrides are being
+        checked here.
+    dispatcher : callable
+        Function that when called like `dispatcher(*args, **kwargs)` returns an
+        iterable of relevant argument to check to for __array_function__
+        attributes.
+    args : tuple
+        Arbitrary positional arguments originally passed into api_func.
+    kwargs : tuple
+        Arbitrary keyword arguments originally passed into api_func.
+
+    Returns
+    -------
+    Result from calling `implementation_func()` or an `__array_function__`
+    method, as appropriate.
+
+    Raises
+    ------
+    TypeError : if no implementation is found.
+    """
+
+    # Collect array-like arguments.
+    relevant_arguments = dispatcher(*args, **kwargs)
+
+    # Check for __array_function__ methods.
+    types, overloaded_args = get_overloaded_types_and_args(
+        relevant_arguments)
+
+    # Fast path
+    if not overloaded_args:
+        return implementation_func(*args, **kwargs)
+
+    # Call overrides
     for overloaded_arg in overloaded_args:
         # Note that we're only calling __array_function__ on the *first*
         # occurence of each argument type. This is necessary for reasonable
         # performance with a possibly long list of overloaded arguments, for
         # which each __array_function__ implementation might reasonably need to
         # check all argument types.
-        result = overloaded_arg.__array_function__(func, types, args, kwargs)
+        # api_func is the function exposed in NumPy's public API. We
+        # use it instead of func so __array_function__ implementations
+        # can do equality/identity comparisons.
+        result = overloaded_arg.__array_function__(
+            api_func, types, args, kwargs)
 
         if result is not NotImplemented:
             return result
 
     raise TypeError('no implementation found for {} on types that implement '
                     '__array_function__: {}'
-                    .format(func, list(map(type, overloaded_args))))
+                    .format(api_func, list(map(type, overloaded_args))))
```

Comment:
Only noticing this now, but wouldn't it make sense to just print `types` in the error message? This will include `ndarray` if an array was present, but I think that is logically correct: in a way, its `__array_function__` has been (implicitly) tried, and ti cannot handle it.

Reply:
I'm would lean against this since we don't actually call `ndarray.__array_function__`.

Reply:
We only skip it because we know its answer! But not a big deal either way.

Reply:
We also skip it because `ndarray.__array_function__` calls the public API function rather than the implementation function, so we need to avoid the infinite loop :)